### DISCUSSION TITLE
Highlight brackets of dictionaries and lists in Python output

### DISCUSSION THREAD

**Discussion by benho93** [facelessuser/BracketHighlighter#614] [Q&A] [answered]
Hello,

I would like to be able to highlight dictionaries and lists in the output of my Python scripts. 

Any idea on how to make this happen? 

I have played around a bit with `"search_in_widgets": true` but that doesn't seem to change much. 

Adding the following to the user settings didn't work, either.

```
{
    "name": "Plain Text",
    "scope": "text.plain",
    "brackets": [
        ["(", ")", "punctuation.definition.group"],
        ["[", "]", "punctuation.definition.array"],
        ["{", "}", "punctuation.definition.dict"],
        ["<", ">", "punctuation.definition.generic"],
        ["\"", "\"", "punctuation.definition.string"],
        ["'", "'", "punctuation.definition.string"],
        ["`", "`", "punctuation.definition.string"],
        ["\\\\", "\\\\", "punctuation.definition.escape"]
    ]
}

```

Thank you!

**Comment by facelessuser** [comment, marked as answer]
What you are showing isn't really how the rules work. If you have `search_in_widgets` enabled, that is at least the first part. There aren't profiles for a language type, but you either include language types for a bracket or exclude language types.

We also have two types of brackets: scope brackets and normal brackets. You don't necessarily need to know how those work, just know that there are two different kinds and that you would alter them using different options.

What we *don't* do is modify the default list of `brackets` or `scope_brackets`. You'll use `user_brackets` and `user_scope_brackets`. You must at least specify the `name`, and then you can override any configuration at the root level. so if I wanted to change the default square bracket rule to not be excluded from plain text, then I override that bracket's `language_list` so that it no longer excludes `Plain Text`:

```js
    "user_brackets": [
        {
            "name": "square",
            "language_list": ["Hex", "RTF"],
        },
        // More bracket rules ...
    ]
```

I may or may not make that clear enough in the docs. I'll take a look sometime soon and see if I need to revise things.

**Comment by benho93** [comment]
Thank you!

```
"user_brackets": [
        {
            "name": "square",
            "language_list": ["Hex", "RTF"],
        },
        {
            "name": "curly",
            "language_list": ["Hex", "RTF"],
        },
    ],
```

This did the trick for me.

However, I would like to do the same to double_quotes and single_quotes. Why does the following not work?

```
"user_scope_brackets": [
        {
            "name": "single_quote",
            "language_list": ["Hex", "RTF"],
        },
        {
            "name": "double_quote",
            "language_list": ["Hex", "RTF"],
        },

        // More bracket rules ...
    ]
```

**Reply by facelessuser** [reply]
It's because they are scope brackets, meaning they depend on the language scoping.

Quotes have identical openings and closings. BH works on a sliding window withing the buffer for performance, so it doesn't start at the very beginning of the file, so it depends on scoping for certain brackets like quotes. If it didn't, it would have no idea if it was looking at a beginning quote or an end quote.

That is why they don't work in plain text. Plain text has no scoping, as a matter of fact, plain text has no real structure because anything goes in plain text. Languages like Python have strict rules, and with syntax highlighting enabled, we can get the scopes to give us context. Now, I know that you are displaying Python output in your plain text buffer, but without scoping, "scope brackets" won't work.
